dissect/target/filesystems/nfs.py

```diff
@@ -0,0 +1,168 @@
+from __future__ import annotations
+
+import stat
+from typing import BinaryIO, Callable, Iterator
+
+from dissect.util.stream import AlignedStream
+
+from dissect.target.filesystem import Filesystem, FilesystemEntry
+from dissect.target.helpers import fsutil
+from dissect.target.helpers.nfs.nfs3 import (
+    EntryPlus3,
+    FileAttributes3,
+    FileHandle3,
+    FileType3,
+)
+from dissect.target.helpers.nfs.nfs_client import Client
+
+ClientFactory = Callable[[], Client]
+
+
+class NfsFilesystem(Filesystem):
+    """Filesystem implementation of a NFS share
+
+    The client is lazily constructed
+    """
+
+    __type__ = "nfs"
+
+    def __init__(self, client_factory: ClientFactory, root_handle: FileHandle3):
+        super().__init__()
+        self._client_factory = client_factory
+        self._backing_client = None
+        self._root_handle = root_handle
+
+    @property
+    def _client(self) -> Client:
+        if self._backing_client is None:
+            self._backing_client = self._client_factory()
+        return self._backing_client
+
+    @staticmethod
+    def detect(_: BinaryIO) -> bool:
+        raise TypeError("Detect is not allowed on a NfsFilesystem class")  # :/
+
+    def get(self, path: str) -> NfsFilesystemEntry:
+        path = fsutil.normalize(path, self.alt_separator).strip("/")
+
+        if not path:
+            return NfsFilesystemEntry(self, "/", self._root_handle)
+
+        current_handle = self._root_handle
+        for segment in path.split("/"):
+            result = self._client.lookup(segment, current_handle)
+            current_handle = result.object
+
+        return NfsFilesystemEntry(self, path, result.object, result.obj_attributes)
+
+
+class NfsFilesystemEntry(FilesystemEntry):
```

Comment:
```suggestion
class NfsFilesystemEntry(FilesystemEntry):
    fs: NfsFilesystem
    entry: FileHandle3
```

Reply:
Can't, `FileSystemEntry` upcasts `fs` to `Filesystem`.

Reply:
What do you mean with "can't"? As in that a real type checker isn't satisfied with this? This works fine in e.g. `SmbFilesystem`.

Reply:
You're right, I misread that as being class variables, but they are actually typehints on instance variables
(https://peps.python.org/pep-0526/#class-and-instance-variable-annotations)

Reply:
(I still don´t get why `fs` and `entry` are defined on the `FilesystemEntry` though. `entry` is not used at all in `FilesystemEntry`, and `fs` primarily for the `alt_sepaparator`)

Reply:
It's for interactive usage in IPython shells, it gives a user access to the raw underlying filesystem object under a standardized attribute name.